core/src/main/scala/org/apache/spark/scheduler/BlacklistTracker.scala

```diff
@@ -0,0 +1,247 @@
+            val failureStatus = new FailureStatus(
+              clock.getTimeMillis(), info.host, failedTasks)
+            executorIdToFailureStatus.update(executorId, failureStatus)
+        }
+        invalidAllCache()
+    }
+  }
+
+  /** remove the executorId from executorIdToFailureStatus */
+  def removeFailureExecutors(executorId: String) : Unit = synchronized {
+    executorIdToFailureStatus.remove(executorId)
+    invalidAllCache()
+  }
+
+  /**
+   * remove the failure record related to given taskId from executorIdToFailureStatus. If the
+   * number of records of given executorId becomes 0, remove the completed executorId.
+   */
+  def removeFailureExecutorsForTaskId(
+      executorId: String,
+      stageId: Int,
+      taskIndex: Int) : Unit = synchronized {
+    val atomTask = BlacklistAtomTask(stageId, taskIndex)
+    executorIdToFailureStatus.get(executorId).map(fs => {
+      fs.numFailuresPerTask.remove(atomTask)
+      if(fs.numFailuresPerTask.isEmpty){
+        executorIdToFailureStatus.remove(executorId)
+      }
+      invalidAllCache()
+    })
+  }
+
+  def executorIsBlacklisted(
+      executorId: String,
+      sched: TaskSchedulerImpl,
+      stageId: Int,
+      taskIndex: Int) : Boolean = {
+
+    executorBlacklist(sched, stageId, taskIndex).contains(executorId)
+  }
+
+  // If the node is in blacklist, all executors allocated on that node will
+  // also be put into  executor blacklist.
+  private def executorsOnBlacklistedNode(
+      sched: TaskSchedulerImpl): Set[String] = {
+      nodeBlacklist().flatMap(sched.getExecutorsAliveOnHost(_)
+        .getOrElse(Set.empty[String])).toSet
+  }
+
+  def executorBlacklist(
+      sched: TaskSchedulerImpl, stageId: Int, taskIndex: Int): Set[String] = synchronized {
+    val atomTask = BlacklistAtomTask(stageId, taskIndex)
+    if (!isBlacklistExecutorCacheValid) {
+      reEvaluateExecutorBlacklistAndUpdateCache(sched, atomTask, clock)
+    } else {
+      getExecutorBlacklistFromCache(atomTask).getOrElse {
+        reEvaluateExecutorBlacklistAndUpdateCache(sched, atomTask, clock)
+      }
+    }
+  }
+
+  private def reEvaluateExecutorBlacklistAndUpdateCache(
+      sched: TaskSchedulerImpl,
+      atomTask: BlacklistAtomTask,
+      clock: Clock): Set[String] = {
+    val executors = executorsOnBlacklistedNode(sched) ++
+      strategy.getExecutorBlacklist(executorIdToFailureStatus, atomTask, clock)
+    updateBlacklistExecutorCache(atomTask, executors)
+    executors
+  }
+
+  // The actual implementation is delegated to strategy
+  def nodeBlacklist(): Set[String] = synchronized {
+    if (isBlacklistNodeCacheValid) {
+      getNodeBlacklistFromCache
+    } else {
+      val nodes = strategy.getNodeBlacklist(executorIdToFailureStatus)
+      updateBlacklistNodeCache(nodes)
+      nodes
+    }
+  }
+}
+
+/**
+ * Hide cache details in this trait to make code clean and avoid operation mistake
+ */
+private[scheduler]  trait BlacklistCache {
```

Comment:
actually ... in the end this trait is only used in one place, its not even tested separately.  So I think you should just move this code inside `BlacklistTracker`.